### PR TITLE
Use TargetTriple::from_path in rustdoc

### PR DESCRIPTION
**Pull request by bjorn3**
This fixes the problem reported in https://github.com/Rust-for-Linux/linux/pull/272 where rustdoc requires the absolute path of a target spec json instead of accepting a relative path like rustc.

### CODE REVIEW

**Comment by bjorn3** [comment]
```
  /usr/bin/git -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=2 origin +d20bfe8ac965fce0c261b2fb0771fbf29324d906:refs/remotes/pull/85361/merge
  remote: Internal Server Error.
  remote: 
  Error: fatal: unable to access 'https://github.com/rust-lang/rust/': The requested URL returned error: 500
  The process '/usr/bin/git' failed with exit code 128
```

**Comment by jyn514** [comment]
Can you add a test for the new behavior (maybe a run-make test)?

Where is this logic for rustc, is it possible to reuse it?

**Comment by bjorn3** [comment]
The logic for rustc is at https://github.com/rust-lang/rust/blob/544d124b81fb54e1904b6cd53de9bf92b6b3d0b9/compiler/rustc_session/src/config.rs#L1510-L1521

**Comment by jyn514** [comment]
Great. Can you make that function public and use it in rustdoc instead of duplicating it?

**Comment by jyn514** [comment]
@bjorn3 you made the function public, but didn't use it in rustdoc. Also it would be nice to have the test.

**Comment by bjorn3** [comment]
I did `git add compiler`, which doesn't include `src/librustdoc`. How should I test it? Should I compile the real libcore or use a dummy `#![no_core]` lib as dependency?

**Comment by jyn514** [comment]
The error in https://github.com/Rust-for-Linux/linux/pull/258#issuecomment-835669054 confuses me (that is the error this fixes, right?) but if you can write dummy no_core test that fails before and passes after, that sounds great :)

**Comment by bjorn3** [comment]
> that is the error this fixes, right?

Yes

> The error in Rust-for-Linux/linux#258 (comment) confuses me

The problem is that when using rustc the full canonicalized path of the target spec json would be used by rustc and thus end up in the metadata of libcore and other dependencies, but when trying to use rustdoc, only the literal value of `--target` is used, which results in rustc rejecting the metadata due to a different target value.

> but if you can write dummy no_core test that fails before and passes after, that sounds great :)

:+1:

File: src/test/run-make-fulldeps/rustdoc-target-spec-json-path/Makefile [hunk 1]
```diff
@@ -0,0 +1,9 @@
+-include ../tools.mk
```
Comment (jyn514) [inline src/test/run-make-fulldeps/rustdoc-target-spec-json-path/Makefile:1]: nit: can you use `include` instead so it gives an error if the file is missing?

**Review by jyn514 — APPROVED** [review]
r=me if you confirmed the test fails before this PR (I can't check right now).

@bors delegate=bjorn3

File: src/test/run-make-fulldeps/rustdoc-target-spec-json-path/Makefile (hunk 1 above)
Comment (bjorn3) [reply src/test/run-make-fulldeps/rustdoc-target-spec-json-path/Makefile:1]: Sure, I just copied another rustdoc test.

**Comment by bjorn3** [comment]
I manually ran the commands in the test. It failed with a nightly rustc/rustdoc.

@bors r=jyn514

**Comment by bjorn3** [comment]
Fixed the test.

@bors r=jyn514

**Pull request merged**